### PR TITLE
MAINT: Remove special cases in np.unique

### PR DESCRIPTION
**Pull request by eric-wieser**
Inspired by looking at gh-9531

### CODE REVIEW

File: numpy/lib/arraysetops.py [hunk 1]
```diff
@@ -259,45 +266,31 @@ def _unique1d(ar, return_index=False, return_inverse=False,
...
-            iflag = np.cumsum(flag) - 1
-            inv_idx = np.empty(ar.shape, dtype=np.intp)
-            inv_idx[perm] = iflag
-            ret += (inv_idx,)
-        if return_counts:
-            idx = np.concatenate(np.nonzero(flag) + ([ar.size],))
-            ret += (np.diff(idx),)
+    mask = np.empty(aux.shape, dtype=np.bool_)
+    mask[:1] = True
+    mask[1:] = aux[1:] != aux[:-1]
```
Comment (eric-wieser) [inline numpy/lib/arraysetops.py]: Doing this makes the size match for empty arrays too

Comment (jaimefrio) [reply numpy/lib/arraysetops.py]: Clever! But maybe it's worth adding a comment? The `mask[:1] = True` is a little puzzling until you understand it works with empty arrays.

File: numpy/lib/arraysetops.py [hunk 2]
```diff
@@ -259,45 +268,31 @@ def _unique1d(ar, return_index=False, return_inverse=False,
...
+
+    ret = (aux[mask],)
+    if return_index:
+        ret += (perm[mask],)
+    if return_inverse:
+        imask = np.cumsum(mask) - 1
```
Comment (jaimefrio) [inline numpy/lib/arraysetops.py]: I'm surprised that the `- 1` works for an empty array. But it does work!

**Review by jaimefrio — APPROVED** [review]
Nice refactorings!

**Comment by jaimefrio** [comment]
Thanks Eric!

**Pull request merged**